`wp-env install-path` does not work when offline

### Description

When internet is turned off, you'll get an error trying connect to `api.wordpress.org`.

All I wanted was getting the install path.

### Step-by-step reproduction instructions

1. Turn off internet access
2. Run `wp-env install-path`
3. Get error message

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

I was able to reproduce the problem and will leave an error log.

```
$ npm run wp-env install-path

> gutenberg@16.1.0 wp-env /home/username/projects/gutenberg
> wp-env "install-path"

✖ getaddrinfo EAI_AGAIN api.wordpress.org
RequestError: getaddrinfo EAI_AGAIN api.wordpress.org
    at ClientRequest.<anonymous> (/home/username/projects/gutenberg/node_modules/got/dist/source/core/index.js:970:111)
    at Object.onceWrapper (events.js:520:26)
    at ClientRequest.emit (events.js:412:35)
    at ClientRequest.origin.emit (/home/username/projects/gutenberg/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)
    at TLSSocket.socketErrorListener (_http_client.js:475:9)
    at TLSSocket.emit (events.js:400:28)
    at emitErrorNT (internal/streams/destroy.js:106:8)
    at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    at processTicksAndRejections (internal/process/task_queues.js:82:21)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:71:26) {
  code: 'EAI_AGAIN',
  timings: {
    start: 1688210948625,
    socket: 1688210948625,
    lookup: 1688210958635,
    connect: undefined,
    secureConnect: undefined,
    upload: undefined,
    response: undefined,
    end: undefined,
    error: 1688210958636,
    abort: undefined,
    phases: {
      wait: 0,
      dns: 10010,
      tcp: undefined,
      tls: undefined,
      request: undefined,
      firstByte: undefined,
      download: undefined,
      total: 10011
    }
  }
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gutenberg@16.1.0 wp-env: `wp-env "install-path"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the gutenberg@16.1.0 wp-env script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/username/.npm/_logs/2023-07-01T11_29_18_659Z-debug.log
```

--- a/packages/env/CHANGELOG.md
+++ b/packages/env/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### Enhancement
+
+-   `wp-env` now works offline after the environment has been created. Note that many `wp-env` configuration changes involve internet connectivity and may not work in offline mode. [#53547](https://github.com/WordPress/gutenberg/pull/53547)
+
 ## 8.10.0 (2023-10-18)
 
 ### Bug Fix

--- a/packages/env/lib/cache.js
+++ b/packages/env/lib/cache.js
@@ -40,6 +40,9 @@ async function didCacheChange( key, value, options ) {
  * @param {WPEnvCacheOptions} options Parsing options
  */
 async function setCache( key, value, options ) {
+	// Make sure the cache directory exists.
+	await fs.mkdir( options.workDirectoryPath, { recursive: true } );
+
 	const existingCache = await getCacheFile( options );
 	existingCache[ key ] = value;
 

--- a/packages/env/lib/commands/start.js
+++ b/packages/env/lib/commands/start.js
@@ -28,6 +28,7 @@ const {
 	configureWordPress,
 	setupWordPressDirectories,
 	readWordPressVersion,
+	canAccessWPORG,
 } = require( '../wordpress' );
 const { didCacheChange, setCache } = require( '../cache' );
 const md5 = require( '../md5' );
@@ -77,15 +78,23 @@ module.exports = async function start( {
 	const configHash = md5( config );
 	const { workDirectoryPath, dockerComposeConfigPath } = config;
 	const shouldConfigureWp =
-		update ||
-		( await didCacheChange( CONFIG_CACHE_KEY, configHash, {
-			workDirectoryPath,
-		} ) );
+		( update ||
+			( await didCacheChange( CONFIG_CACHE_KEY, configHash, {
+				workDirectoryPath,
+			} ) ) ) &&
+		// Don't reconfigure everything when we can't connect to the internet because
+		// the majority of update tasks involve connecting to the internet. (Such
+		// as downloading sources and pulling docker images.)
+		( await canAccessWPORG() );
 
 	const dockerComposeConfig = {
 		config: dockerComposeConfigPath,
 		log: config.debug,
 	};
+
+	if ( ! ( await canAccessWPORG() ) ) {
+		spinner.info( 'wp-env is offline' );
+	}
 
 	/**
 	 * If the Docker image is already running and the `wp-env` files have been

--- a/packages/env/lib/config/parse-config.js
+++ b/packages/env/lib/config/parse-config.js
@@ -534,7 +534,7 @@ async function parseEnvironmentConfig(
 async function parseCoreSource( coreSource, options ) {
 	// An empty source means we should use the latest version of WordPress.
 	if ( ! coreSource ) {
-		const wpVersion = await getLatestWordPressVersion();
+		const wpVersion = await getLatestWordPressVersion( options );
 		if ( ! wpVersion ) {
 			throw new ValidationError(
 				'Could not find the latest WordPress version. There may be a network issue.'

--- a/packages/env/lib/config/test/config-integration.js
+++ b/packages/env/lib/config/test/config-integration.js
@@ -15,6 +15,8 @@ jest.mock( 'fs', () => ( {
 	promises: {
 		readFile: jest.fn(),
 		stat: jest.fn().mockResolvedValue( true ),
+		mkdir: jest.fn(),
+		writeFile: jest.fn(),
 	},
 } ) );
 

--- a/packages/env/lib/test/cache.js
+++ b/packages/env/lib/test/cache.js
@@ -18,6 +18,7 @@ jest.mock( 'fs', () => ( {
 	promises: {
 		readFile: jest.fn(),
 		writeFile: jest.fn(),
+		mkdir: jest.fn(),
 	},
 } ) );
 

--- a/packages/env/lib/wordpress.js
+++ b/packages/env/lib/wordpress.js
@@ -254,8 +254,14 @@ async function readWordPressVersion( coreSource, spinner, debug ) {
  *
  * @return {boolean} True if we can connect to WordPress.org, false otherwise.
  */
+let IS_OFFLINE;
 async function canAccessWPORG() {
-	return !! ( await dns.resolve( 'WordPress.org' ).catch( () => {} ) );
+	// Avoid situations where some parts of the code think we're offline and others don't.
+	if ( IS_OFFLINE !== undefined ) {
+		return IS_OFFLINE;
+	}
+	IS_OFFLINE = !! ( await dns.resolve( 'WordPress.org' ).catch( () => {} ) );
+	return IS_OFFLINE;
 }
 
 /**

--- a/packages/env/lib/wordpress.js
+++ b/packages/env/lib/wordpress.js
@@ -7,11 +7,17 @@ const util = require( 'util' );
 const fs = require( 'fs' ).promises;
 const path = require( 'path' );
 const got = require( 'got' );
+const dns = require( 'dns' ).promises;
 
 /**
  * Promisified dependencies
  */
 const copyDir = util.promisify( require( 'copy-dir' ) );
+
+/**
+ * Internal dependencies
+ */
+const { getCache, setCache } = require( './cache' );
 
 /**
  * @typedef {import('./config').WPConfig} WPConfig
@@ -244,16 +250,45 @@ async function readWordPressVersion( coreSource, spinner, debug ) {
 }
 
 /**
+ * Basically a quick check to see if we can connect to the internet.
+ *
+ * @return {boolean} True if we can connect to WordPress.org, false otherwise.
+ */
+async function canAccessWPORG() {
+	return !! ( await dns.resolve( 'WordPress.org' ).catch( () => {} ) );
+}
+
+/**
  * Returns the latest stable version of WordPress by requesting the stable-check
  * endpoint on WordPress.org.
  *
+ * @param {Object} options an object with cacheDirectoryPath set to the path to the cache directory in ~/.wp-env.
  * @return {string} The latest stable version of WordPress, like "6.0.1"
  */
 let CACHED_WP_VERSION;
-async function getLatestWordPressVersion() {
+async function getLatestWordPressVersion( options ) {
 	// Avoid extra network requests.
 	if ( CACHED_WP_VERSION ) {
 		return CACHED_WP_VERSION;
+	}
+
+	const cacheOptions = {
+		workDirectoryPath: options.cacheDirectoryPath,
+	};
+
+	// When we can't connect to the internet, we don't want to break wp-env or
+	// wait for the stable-check result to timeout.
+	if ( ! ( await canAccessWPORG() ) ) {
+		const latestVersion = await getCache(
+			'latestWordPressVersion',
+			cacheOptions
+		);
+		if ( ! latestVersion ) {
+			throw new Error(
+				'Could not find the current WordPress version in the cache and the network is not available.'
+			);
+		}
+		return latestVersion;
 	}
 
 	const versions = await got(
@@ -263,6 +298,7 @@ async function getLatestWordPressVersion() {
 	for ( const [ version, status ] of Object.entries( versions ) ) {
 		if ( status === 'latest' ) {
 			CACHED_WP_VERSION = version;
+			await setCache( 'latestWordPressVersion', version, cacheOptions );
 			return version;
 		}
 	}

--- a/packages/env/lib/wordpress.js
+++ b/packages/env/lib/wordpress.js
@@ -311,5 +311,6 @@ module.exports = {
 	resetDatabase,
 	setupWordPressDirectories,
 	readWordPressVersion,
+	canAccessWPORG,
 	getLatestWordPressVersion,
 };

